src/PortAudio.jl

```diff
@@ -218,29 +187,62 @@ function PortAudioStream(
     inchans = 2,
     outchans = 2;
     eltype = Float32,
-    samplerate = -1,
+    samplerate = combine_default_sample_rates(
+        inchans,
+        indev.defaultsamplerate,
+        outchans,
+        outdev.defaultsamplerate,
+    ),
     latency = defaultlatency(indev, outdev),
     warn_xruns = false,
     recover_xruns = true,
+    frames_per_buffer = 0,
+    flags = paNoFlag,
+    callback = nothing,
+    user_data = nothing,
+    input_info = nothing,
+    output_info = nothing,
 )
-    if samplerate == -1
-        samplerate = combine_default_sample_rates(
-            inchans,
-            indev.defaultsamplerate,
-            outchans,
-            outdev.defaultsamplerate,
-        )
-    end
-    PortAudioStream{eltype}(
-        indev,
-        outdev,
-        inchans,
-        outchans,
+    inchans = fill_max_channels(inchans, indev.input_bounds)
+    outchans = fill_max_channels(outchans, outdev.output_bounds)
+    pointer_ref = @stderr_as_debug Pa_OpenStream(
+        make_parameters(indev, inchans, eltype, latency, input_info),
+        make_parameters(outdev, outchans, eltype, latency, output_info),
+        samplerate,
+        frames_per_buffer,
+        flags,
+        callback,
+        user_data,
+    )
+    Pa_StartStream(pointer_ref[])
+    this = PortAudioStream{eltype}(
         samplerate,
         latency,
+        pointer_ref,
         warn_xruns,
         recover_xruns,
+        Buffer{eltype}(outdev, outchans),
+        Buffer{eltype}(indev, inchans),
     )
+    # pre-fill the output stream so we're less likely to underrun
+    prefill_output(this.sink)
+    this
+end
+
+function device_by_name(device_name)
```

Comment:
create the error message while we are looping through the devices so we don't have to do it it twice